### ISSUE TITLE
Broken objects after trying to import/remove flight model demo objects

### ISSUE THREAD

**Issue by mydoghasfleas** [abapGit/abapGit#6924]
On our on-prem S/4 HANA 2023 system, I tried to import the 2023 branch of the flight model demo (https://github.com/SAP-samples/abap-platform-refscen-flight) from an exported zip file, as our system is blocked by the firewall from accessing github.com.

I was using the "1.121.0 - Standalone Version" ABAP Git program that came with the S/4 system.

The initial import did not manage to bring in all the objects, e.g. NRO (same as reported here https://github.com/SAP-samples/abap-platform-refscen-flight/issues/17).

I then tried to Uninstall the package, but this left more object that could not be deleted.

Then, in my desperation, I tried to import the version of the flight model objects that comes with the "ABAP RESTful Application Programming Model" book from SAP PRESS (which is the reason I am importing the flight model in the first place), thinking it will perhaps override the objects, but this seems just to have made matters worse.

I have spent some hours trying to manually delete all the objects created in the /DMO/ namespace, but it seems now I have a number of object entries that are in the state described in note 1156585. Deleting a seemingly empty package is prevented by the assignment of objects which seem to still have a TADIR entry but no actual other presence.

I'm hoping that, in future, abapGit would have some kind of recovery or emergency utility to sort out these inconsistencies or, better yet, prevent them altogether.

A this stage, however, I am a bit wary of using abapGit for large and complex deployments, as I don't think it is mature enough to handle it.

![image](https://github.com/abapGit/abapGit/assets/125179869/59d193c8-87f8-4ed8-83de-ca2f6035ac12)

**Comment by larshp** [comment]
> Deleting a seemingly empty package is prevented by the assignment of objects which seem to still have a TADIR entry but no actual other presence

thats SAP behavior defined by SAP, we cannot do anything about that

SAP provides the sflight using latest features, seemingly without testing, I recommend to test things work before putting it into the public

SAP recently released the ABAP cloud trial for 2022, until then the community has only been able to test 1909

as for the sflight, its not needed in the productive system, so I recommend putting it into a non-transported package

Pull requests welcome with fixes

**Comment by larshp** [comment]
from an abapGit perspective I think we would like to test everything as often as possible, but as open source we dont have any license to run any SAP product and provide services to the internet

also the budget for running it is $0, so the project does rely on users to help testing on different versions

**Comment by mydoghasfleas** [comment]
> also the budget for running it is $0, so the project does rely on users to help testing on different versions

That is interesting. I was under obviously laboring under the misapprehension that SAP is somehow funding it, at least by employing the maintainers of abapGit to work on it 🤔

Considering they shipped it with our S/4 HANA system, it certainly appears they endorse it. Least they could do is provide some funding!

![image](https://github.com/abapGit/abapGit/assets/125179869/7817d06a-1ede-4aed-8445-e94f3277b022)

Anyway, I think I will reach out to the sflight maintainers and explain the predicament. Perhaps they can also provide their releases as transport packages. Thanks!

**Comment by larshp** [comment]
> That is interesting. I was under obviously laboring under the misapprehension that SAP is somehow funding it, at least by employing the maintainers of abapGit to work on it 

I dont think SAP has provided a free cup of coffee to any of the maintainers ☕ 

transport packages: its possible, but doesnt provide transparency as to what is changed and whats part of the transport, and makes it impossible for users to contribute back, so thats also a loose-loose situation IMHO

**Comment by larshp** [comment]
https://github.com/SAP-samples/abap-platform-refscen-flight/issues/34 opened

**Comment by mydoghasfleas** [comment]
> I dont think SAP has provided a free cup of coffee to any of the maintainers ☕

Yet their endorsement of abapGit stretches even to officially sanctioned SAP tutorials (like this one: https://developers.sap.com/tutorials/abap-environment-abapgit-transfer.html).

IMO, they are playing a very dangerous game! 🔥🔥🔥

**Comment by larshp** [comment]
abapGit can be used in mostly any way by everyone, https://github.com/abapGit/abapGit/blob/main/LICENSE

**Comment by pokrakam** [comment]
> Considering they shipped it with our S/4 HANA system, it certainly appears they endorse it. The least they could do is provide some funding!

They do, but `endorse <> support`. There was an attempt by SAP to officialise and support it but that failed miserably because they thought it would be a good idea to put it their version into an own namespace that nobody else could modify. I can't remember the name (can't immediately find it, but I vaguely recall something about Aarhus or Aalborg or some Danish city?). 

Also, regarding:
> this stage, however, I am a bit wary of using abapGit for large and complex deployments, as I don't think it is mature enough to handle it.

Lars already explained the resources the contributors have available (2-3 major versions behind on average), but I'd like to also point out the modular nature of the object de/serialisers. 
abapGit consists of a core application and a bunch of distinct modules to handle each object type. Programs and classes have been around for a long time and are pretty stable. RAP objects are new and under active development by SAP. Until a month ago we only had 1909 with limited RAP available. So when SAP release 2023, it is reasonable to expect that objects under active development by SAP may have some differences in the way they are created/updated.

BTW, I tried uninstalling the flight model on 2022 and ran into similar issues in that the RAP stuff didn't uninstall cleanly. I don't know if this is an issue with abapGit or SAP. At first glance it looks like there may be a dependency to something unsupported and the deletion doesn't fail cleanly. I'll try to have a further look into this and raise a more specific issue if I can identify a clear scenario.

**Comment by mydoghasfleas** [comment]
(I had another comment I was writing, but I'm glad to only post it now after @pokrakam's comment).

Sorry, no offense meant. I think I should clarify what I mean:

It has been my experience that SAP has a history of taking from the developer community and giving little or nothing back. An example would be that there is, on the one hand, a seeming move to encourage innovation in the developer community, but not providing any means to do so. For instance, for a long time, it was impossible to get a trial ABAP system for personal use and even now it has a very restrictive license when, instead, SAP could choose to make it easy for developers to have a system to play with.

Moreover, while SAP makes some things open source, like UI5, some of the VS Code plugins built around it are not.

The big one of course is the ABAP kernel. Understandably, on one hand, much of their competitive advantage is tied up in ABAP, as is a lot of their intellectual property, so we are unlikely to see that ever.

But SAP, in my opinion, does very little overall to help the developer community, from which they benefit a great deal. That there are people who, in the face of such conditions, are still prepared to build on top of their platforms in an open source manner, is quite mind-boggling to me.

I mean, thank you for all your efforts, and I really appreciate and applaud what you are doing, but it seems kind of insane.

**Comment by sbcgua** [comment]
> this stage, however, I am a bit wary of using abapGit for large and complex deployments, as I don't think it is mature enough to handle it.

I'm doing this for almost 10 years by now with our 2 products.
- ~10 clients (read: environments), ECC and S4
- One of the products is a complex Tax reporting tool (kind of business critical)
- 800+ abap object
- DDIC+PROG+CLAS+FUGR+XSLT+MIME+BADI obj types (and more)
- 100k+ lines of code

No major issues so far. Though it was not 100% smooth in early days ;) And I still have some deplyment-centric wishes ... and contributing them ... and planning to continue doing so (though a bit tough time-wise lately).

**Comment by pokrakam** [comment]
Found it! It was called [Project Odense](https://community.sap.com/t5/technology-blogs-by-sap/inside-steampunk-vital-parts-of-steampunk-s-abapgit/ba-p/13468386) (hah, I was right about a Danish city, not a complete goldfish memory yet). 

It was a nice idea and the blog was full of optimism, but not sure what SAP expected us to do with a locked down tool though. They should have open-sourced it. Perhaps unsurprisingly it has recently been [deprecated](https://github.com/SAP-archive/project-odense), which is a shame.

**Comment by gregorwolf** [comment]
Hi @mydoghasfleas,

beside the issue [Provide transport packages as releases #35](https://github.com/SAP-samples/abap-platform-refscen-flight/issues/35) that you have created I would suggest that you also file one at https://github.com/SAP-samples/abap-platform-refscen-flight to track the issue the import causes there. Because I think it should be tracked and solved by SAP there.

**Comment by mydoghasfleas** [comment]
After eventually managing to remove the objects from the system, I am repeating the process. This time, again, there are issues (the same ones I started with originally) but I will log a separate issue for that.
